feat(fmt): `Yul` support

rkrasiuk opened the pull request:
## Solution

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

- [x] `YulBlock`

`YulStatement`s:
- [x] `Assign`
- [x] `VariableDeclaration`
- [x] `If`
- [x] `For`
- [x] `Switch`
- [x] `Leave`
- [x] `Break`
- [x] `Continue`
- [x] `Block`
- [x] `FunctionDefinition`
- [x] `FunctionCall`

`YulExpression`s:
- [x] `BoolLiteral`
- [x] `NumberLiteral`
- [x] `HexNumberLiteral`
- [x] `HexStringLiteral`
- [x] `StringLiteral`
- [x] `Variable`
- [x] `FunctionCall`
- [x] `SuffixAccess`

gakonst commented:
@rkrasiuk https://github.com/brockelmore/solvm this repo by @brockelmore has a lot of Yul inside which could be useful to make testcases out of

mattsse commented on fmt/testdata/Yul/fmt.sol:
```diff
@@ -0,0 +1,184 @@
+            function empty() {}
+
+            function functionThatReturnsSevenValuesAndCanBeUsedInAssignment() ->
+                v1,
+                v2,
+                v3,
+                v4,
+                v5,
+                v6,
+                v7
```
this is probably a bit opinionated, but perhaps the nicest variant,

should the `->` also be put on the next line so that we have
`-> v1,`
?

rkrasiuk replied:
we can do that. should the return variable names be at the same indentation level as `->`?

mattsse replied:
I would suggest

```soliditiy
..... -> v1,
.........v2,
```

so params same level and the arrow just before the first param in the list, imo this helps communicating that this list is return types

rkrasiuk replied:
not sure how i feel about this change just yet since all lines after the first one would be not at `level * indent width`, but at `level * indent width + 4` 🤔, which seems like an arbitrary rule.
wdyt about?
```
....->
........v1,
........v2, 
```

mattsse commented on fmt/testdata/Yul/fmt.sol:
```diff
@@ -0,0 +1,184 @@
+            let x, y := sample2(2, 1)
+
+            let val1, val2, val3, val4, val5, val6, val7
+            val1, val2, val3, val4, val5, val6, val7 :=
```
do we need an indent here?

rkrasiuk replied:
no need since these are two separate statements - first declaration, then assignment

mattsse commented on fmt/testdata/EventDefinition/fmt.sol:
```diff
@@ -21,24 +21,10 @@ contract Events {
-        uint256,
-        uint256,
-        uint256
+        uint256, uint256, uint256, uint256, uint256, uint256, uint256, uint256
```
the wrapped `anonymous` looks a bit lost imo, should `anonymous` events wrap earlier?

rkrasiuk replied:
agree that it looks a bit weird. could you clarify what's you suggestion pls?

rkrasiuk replied:
okay, i will revert to the previous behavior  👍

mattsse reviewed (commented):
this looks pretty good.

one thing to keep in minds is that we eventually want to support customizing some rules, worth having a look at https://rust-lang.github.io/rustfmt/?version=v1.5.1&search= rules/options

good wise this lgtm, but have some style related qs, but maybe these were already discussded

jpopesculian commented on fmt/src/formatter.rs:
```diff
@@ -2856,6 +2845,242 @@ impl<'a, W: Write> Visitor for Formatter<'a, W> {
+    ) -> Result<(), Self::Error> {
+        write_chunk!(self, loc.start(), "assembly")?;
+        if let Some(StringLiteral { loc, string, .. }) = dialect {
+            write_chunk!(self, loc.start(), loc.end(), "\"{string}\"")?;
```
but there's enough places where that needs to be implemented so i'm sure it'll just be its own story

jpopesculian commented on fmt/src/formatter.rs:
```diff
@@ -2856,6 +2845,242 @@ impl<'a, W: Write> Visitor for Formatter<'a, W> {
+        attempt_single_line: bool,
+    ) -> Result<(), Self::Error> {
+        if attempt_single_line && statements.len() == 1 {
+            let chunk = self.chunked(loc.start(), Some(loc.end()), |fmt| {
```
i'm worried because we're not doing this in a temporary buffer, this might alter the comments. e.g.

```
assembly { statement /* a really long comment */ }
```

rkrasiuk replied:
just wrote a test for this, if the comment exceeds the max line length, then it ends up multilined. the reason for that is that the block is treated like a chunk and it wouldn't fit on one line

jpopesculian commented on fmt/src/visit.rs:
```diff
@@ -318,6 +318,78 @@ pub trait Visitor {
+        self.visit_source(loc)
+    }
+
+    fn visit_yul_break(&mut self, loc: Loc) -> Result<(), Self::Error> {
```
since this does the same thing, I think we can just use `visit_break` and `visit_continue`

jpopesculian commented on fmt/src/formatter.rs:
```diff
@@ -2856,6 +2845,242 @@ impl<'a, W: Write> Visitor for Formatter<'a, W> {
+        })
+    }
+
+    fn visit_yul_var_declaration(
```
cam we not reuse the stuff from `visit_yul_assignment`

jpopesculian reviewed (commented):
Damn, you're getting fast at this haha :fire: Looks great! I just left a bunch of comments when things popped into my mind, but other than maybe some uncovered test cases, nothing really important.

jpopesculian commented on fmt/src/formatter.rs:
```diff
@@ -1188,26 +1188,116 @@ impl<'a, W: Write> Formatter<'a, W> {
     }
 
-    fn visit_ident_list(
+    fn visit_list<T>(
```
kinda crazy how much use you get out of this bad boy 💯

rkrasiuk commented on fmt/src/formatter.rs:
```diff
@@ -1188,26 +1188,116 @@ impl<'a, W: Write> Formatter<'a, W> {
-                )?;
-                let multiline = fmt.are_chunks_separated_multiline("{}", &args, ", ")?;
+        items: &mut Vec<T>,
+        offsets: (Option<usize>, Option<usize>),
```
agreed

rkrasiuk commented on fmt/src/formatter.rs:
```diff
@@ -1188,26 +1188,116 @@ impl<'a, W: Write> Formatter<'a, W> {
+        T: LineOfCode + Visitable,
+        M: FnMut(&mut Self, &'b mut Box<T>) -> Result<Option<(&'b mut Box<T>, &'b mut Identifier)>>,
+    {
+        let chunk_member_access = |fmt: &mut Self, ident: &mut Identifier, expr: &mut Box<T>| {
```
tbh, this was the easiest way to deal with the references. the `next` here is only the expression, but we need to return both expression & identifier

mattsse reviewed (approved):
lgtm, pending merging conflict

rkrasiuk commented:
@mattsse thanks, there is one more comment issue in blocks pointed out by @jpopesculian I wanna address

rkrasiuk commented:
@mattsse i've implemented the event formatting in a way that it does not leave the `anonymous` stranded.

regarding the block line comments, I've moved it to [a separate issue](https://github.com/foundry-rs/foundry/issues/2410) since the behavior predates this PR. i believe this one is ready for merge

mattsse reviewed (approved):
good stuff!
